File: src/evai/command/BuildCommands.java
package evai.command;

import mindustry.*;
import mindustry.entities.units.*;
import mindustry.gen.*;
import mindustry.world.*;
import mindustry.world.blocks.environment.*;
import arc.struct.*;
import mindustry.game.*;
import java.util.*;

public class BuildCommands implements CommandRegistry.CommandHandler {
    @Override
    public CommandResult handle(String action, String[] args) {
        switch (action) {
            case "build": return build(args);
            case "break": return breakCmd(args);
            default: return new CommandResult(false, "unknown build command: " + action);
        }
    }

    private Block resolveBlock(String name) {
        Seq<Block> blocks = Vars.content.blocks();
        for (int i = 0; i < blocks.size; i++) {
            Block b = blocks.get(i);
            if (b.name.equals(name) || b.name.replace("-", "").equals(name.replace("-", ""))) return b;
        }
        return null;
    }

    private CommandResult build(String[] args) {
        if (args.length == 0) return new CommandResult(false, "usage: build <block> <x> <y> [rotate]");
        Player player = Vars.player;
        if (player == null) return new CommandResult(false, "no player");
        if (args[0].equals("area")) return buildArea(player, args);
        if (args[0].equals("line")) return buildLine(player, args);
        if (args[0].equals("ring")) return buildRing(player, args);
        Block block = resolveBlock(args[0]);
        if (block == null) return new CommandResult(false, "block not found: " + args[0]);
        Integer x = args.length > 1 ? tryParseInt(args[1], null) : null;
        Integer y = args.length > 2 ? tryParseInt(args[2], null) : null;
        if (x == null || y == null) return new CommandResult(false, "invalid coordinates");
        int rotate = args.length > 3 ? tryParseInt(args[3], 0) : 0;
        player.unit().addBuild(new BuildPlan(x, y, rotate, block));
        return new CommandResult(true, "Queued build " + block.name);
    }

    private CommandResult buildArea(Player player, String[] args) {
        if (args.length < 6) return new CommandResult(false, "usage: build area <block> <x1> <y1> <x2> <y2>");
        Block block = resolveBlock(args[1]);
        if (block == null) return new CommandResult(false, "block not found: " + args[1]);
        int x1 = tryParseInt(args[2], 0), y1 = tryParseInt(args[3], 0);
        int x2 = tryParseInt(args[4], 0), y2 = tryParseInt(args[5], 0);
        int minX = Math.max(Math.min(x1, x2), 0);
        int maxX = Math.min(Math.max(x1, x2), Vars.world.width() - 1);
        int minY = Math.max(Math.min(y1, y2), 0);
        int maxY = Math.min(Math.max(y1, y2), Vars.world.height() - 1);
        int count = 0;
        for (int x = minX; x <= maxX; x++) {
            for (int y = minY; y <= maxY; y++) {
                player.unit().addBuild(new BuildPlan(x, y, 0, block));
                count++;
            }
        }
        return new CommandResult(true, "Queued " + count + " plans for " + block.name);
    }

    private CommandResult buildLine(Player player, String[] args) {
        if (args.length < 6) return new CommandResult(false, "usage: build line <block> <x1> <y1> <x2> <y2>");
        Block block = resolveBlock(args[1]);
        if (block == null) return new CommandResult(false, "block not found");
        int x1 = tryParseInt(args[2], 0), y1 = tryParseInt(args[3], 0);
        int x2 = tryParseInt(args[4], 0), y2 = tryParseInt(args[5], 0);
        int dx = x2 > x1 ? 1 : (x2 < x1 ? -1 : 0);
        int dy = y2 > y1 ? 1 : (y2 < y1 ? -1 : 0);
        int steps = Math.max(Math.abs(x2 - x1), Math.abs(y2 - y1));
        int count = 0;
        for (int i = 0; i <= steps; i++) {
            int px = x1 + dx * i, py = y1 + dy * i;
            if (px >= 0 && px < Vars.world.width() && py >= 0 && py < Vars.world.height()) {
                player.unit().addBuild(new BuildPlan(px, py, 0, block));
                count++;
            }
        }
        return new CommandResult(true, "Queued " + count + " plans");
    }

    private CommandResult buildRing(Player player, String[] args) {
        if (args.length < 5) return new CommandResult(false, "usage: build ring <block> <cx> <cy> <radius>");
        Block block = resolveBlock(args[1]);
        if (block == null) return new CommandResult(false, "block not found");
        int cx = tryParseInt(args[2], 0), cy = tryParseInt(args[3], 0);
        int r = tryParseInt(args[4], 0);
        int count = 0;
        for (int angle = 0; angle < 360; angle += 15) {
            double rad = Math.toRadians(angle);
            int px = (int) Math.round(cx + r * Math.cos(rad));
            int py = (int) Math.round(cy + r * Math.sin(rad));
            px = Math.max(0, Math.min(Vars.world.width() - 1, px));
            py = Math.max(0, Math.min(Vars.world.height() - 1, py));
            player.unit().addBuild(new BuildPlan(px, py, 0, block));
            count++;
        }
        return new CommandResult(true, "Queued " + count + " plans");
    }

    private CommandResult breakCmd(String[] args) {
        Player player = Vars.player;
        if (player == null) return new CommandResult(false, "no player");
        if (args.length > 0 && args[0].equals("area")) {
            if (args.length < 5) return new CommandResult(false, "usage: break area <x1> <y1> <x2> <y2>");
            int x1 = tryParseInt(args[1], 0), y1 = tryParseInt(args[2], 0);
            int x2 = tryParseInt(args[3], 0), y2 = tryParseInt(args[4], 0);
            int minX = Math.max(Math.min(x1, x2), 0), maxX = Math.min(Math.max(x1, x2), Vars.world.width() - 1);
            int minY = Math.max(Math.min(y1, y2), 0), maxY = Math.min(Math.max(y1, y2), Vars.world.height() - 1);
            int count = 0;
            for (int x = minX; x <= maxX; x++) {
                for (int y = minY; y <= maxY; y++) {
                    Tile tile = Vars.world.tile(x, y);
                    if (tile != null && tile.block() != null && !(tile.block() instanceof AirBlock) && !(tile.block() instanceof Floor)) {
                        player.unit().addBuild(new BuildPlan(x, y));
                        count++;
                    }
                }
            }
            return new CommandResult(true, "Queued " + count + " break plans");
        }
        Integer x = args.length > 0 ? tryParseInt(args[0], null) : null;
        Integer y = args.length > 1 ? tryParseInt(args[1], null) : null;
        if (x == null || y == null) return new CommandResult(false, "usage: break <x> <y>");
        player.unit().addBuild(new BuildPlan(x, y));
        return new CommandResult(true, "Queued break at (" + x + ", " + y + ")");
    }

    private static int tryParseInt(String s, int fallback) {
        try { return Integer.parseInt(s); } catch (NumberFormatException e) { return fallback; }
    }
    private static Integer tryParseInt(String s, Integer fallback) {
        try { return Integer.parseInt(s); } catch (NumberFormatException e) { return fallback; }
    }
}

File: src/evai/command/BuildCommands.kt
package evai.command

import mindustry.*
import mindustry.gen.*
import mindustry.world.*
import evai.command.CommandRegistry.CommandHandler

class BuildCommands : CommandHandler {
    override fun handle(action: String, args: Array<String>): CommandResult = when (action) {
        "build" -> {
            if (args.isEmpty()) return CommandResult(false, "usage: build <block> <x> <y> [rotate], build area/line/ring <block> <args>")
            val player = Vars.player ?: return CommandResult(false, "no player")
            val sub = args[0].lowercase()
            when (sub) {
                "area" -> {
                    if (args.size < 6) return CommandResult(false, "usage: build area <block> <x1> <y1> <x2> <y2>")
                    val block = resolveBlock(args[1]) ?: return CommandResult(false, "block not found: ${args[1]}")
                    val x1 = args[2].toIntOrNull() ?: return CommandResult(false, "invalid x1")
                    val y1 = args[3].toIntOrNull() ?: return CommandResult(false, "invalid y1")
                    val x2 = args[4].toIntOrNull() ?: return CommandResult(false, "invalid x2")
                    val y2 = args[5].toIntOrNull() ?: return CommandResult(false, "invalid y2")
                    val minX = minOf(x1, x2).coerceAtLeast(0)
                    val maxX = maxOf(x1, x2).coerceAtMost(Vars.world.width() - 1)
                    val minY = minOf(y1, y2).coerceAtLeast(0)
                    val maxY = maxOf(y1, y2).coerceAtMost(Vars.world.height() - 1)
                    var count = 0
                    val unit = player.unit()
                    for (x in minX..maxX) {
                        for (y in minY..maxY) {
                            unit.addBuild(BuildPlan(x, y, 0, block))
                            count++
                        }
                    }
                    CommandResult(true, "Queued $count build plans for ${block.name}")
                }
                "line" -> {
                    if (args.size < 6) return CommandResult(false, "usage: build line <block> <x1> <y1> <x2> <y2>")
                    val block = resolveBlock(args[1]) ?: return CommandResult(false, "block not found: ${args[1]}")
                    val x1 = args[2].toIntOrNull() ?: return CommandResult(false, "invalid x1")
                    val y1 = args[3].toIntOrNull() ?: return CommandResult(false, "invalid y1")
                    val x2 = args[4].toIntOrNull() ?: return CommandResult(false, "invalid x2")
                    val y2 = args[5].toIntOrNull() ?: return CommandResult(false, "invalid y2")
                    val dx = if (x2 > x1) 1 else if (x2 < x1) -1 else 0
                    val dy = if (y2 > y1) 1 else if (y2 < y1) -1 else 0
                    val steps = maxOf(kotlin.math.abs(x2 - x1), kotlin.math.abs(y2 - y1))
                    var count = 0
                    val unit = player.unit()
                    for (i in 0..steps) {
                        val x = x1 + dx * i
                        val y = y1 + dy * i
                        if (x in 0 until Vars.world.width() && y in 0 until Vars.world.height()) {
                            unit.addBuild(BuildPlan(x, y, 0, block))
                            count++
                        }
                    }
                    CommandResult(true, "Queued $count build plans for line of ${block.name}")
                }
                "ring" -> {
                    if (args.size < 5) return CommandResult(false, "usage: build ring <block> <cx> <cy> <radius>")
                    val block = resolveBlock(args[1]) ?: return CommandResult(false, "block not found: ${args[1]}")
                    val cx = args[2].toIntOrNull() ?: return CommandResult(false, "invalid cx")
                    val cy = args[3].toIntOrNull() ?: return CommandResult(false, "invalid cy")
                    val r = args[4].toIntOrNull() ?: return CommandResult(false, "invalid radius")
                    var count = 0
                    val unit = player.unit()
                    for (angle in 0 until 360 step 15) {
                        val rad = Math.toRadians(angle.toDouble())
                        val x = (cx + r * kotlin.math.cos(rad)).toInt().coerceIn(0, Vars.world.width() - 1)
                        val y = (cy + r * kotlin.math.sin(rad)).toInt().coerceIn(0, Vars.world.height() - 1)
                        unit.addBuild(BuildPlan(x, y, 0, block))
                        count++
                    }
                    CommandResult(true, "Queued $count build plans for ring of ${block.name}")
                }
                "queue" -> {
                    CommandResult(false, "build queue not yet implemented")
                }
                else -> {
                    val block = resolveBlock(args[0]) ?: return CommandResult(false, "block not found: ${args[0]}")
                    val x = args.getOrNull(1)?.toIntOrNull() ?: return CommandResult(false, "invalid x")
                    val y = args.getOrNull(2)?.toIntOrNull() ?: return CommandResult(false, "invalid y")
                    val rotate = args.getOrNull(3)?.toIntOrNull() ?: 0
                    val unit = player.unit()
                    unit.addBuild(BuildPlan(x, y, rotate, block))
                    CommandResult(true, "Queued build ${block.name} at ($x, $y)")
                }
            }
        }
        "break" -> {
            val player = Vars.player ?: return CommandResult(false, "no player")
            if (args.isNotEmpty() && args[0] == "area") {
                if (args.size < 5) return CommandResult(false, "usage: break area <x1> <y1> <x2> <y2>")
                val x1 = args[1].toIntOrNull() ?: return CommandResult(false, "invalid x1")
                val y1 = args[2].toIntOrNull() ?: return CommandResult(false, "invalid y1")
                val x2 = args[3].toIntOrNull() ?: return CommandResult(false, "invalid x2")
                val y2 = args[4].toIntOrNull() ?: return CommandResult(false, "invalid y2")
                val minX = minOf(x1, x2).coerceAtLeast(0)
                val maxX = maxOf(x1, x2).coerceAtMost(Vars.world.width() - 1)
                val minY = minOf(y1, y2).coerceAtLeast(0)
                val maxY = maxOf(y1, y2).coerceAtMost(Vars.world.height() - 1)
                var count = 0
                val unit = player.unit()
                for (x in minX..maxX) {
                    for (y in minY..maxY) {
                        unit.addBuild(BuildPlan(x, y, 0, Block.air, true))
                        count++
                    }
                }
                CommandResult(true, "Queued $count break plans")
            } else {
                val x = args.getOrNull(0)?.toIntOrNull() ?: return CommandResult(false, "usage: break <x> <y>, break area <x1> <y1> <x2> <y2>")
                val y = args.getOrNull(1)?.toIntOrNull() ?: return CommandResult(false, "usage: break <x> <y>")
                val unit = player.unit()
                unit.addBuild(BuildPlan(x, y, 0, Block.air, true))
                CommandResult(true, "Queued break at ($x, $y)")
            }
        }
        else -> CommandResult(false, "unknown build command: $action")
    }

    private fun resolveBlock(name: String): Block? {
        return Vars.content.blocks().find { it.name == name || it.name.endsWith("-$name") || it.name == name.lowercase().replace(" ", "-") }
    }
}